### ISSUE TITLE
Changing CDSView filters' attributes does not trigger rendering

### ISSUE THREAD

**Issue by p-himik** [bokeh/bokeh#7273]
Bokeh 0.12.11

```python
from bokeh.io import curdoc
from bokeh.layouts import column
from bokeh.models import CDSView, IndexFilter, ColumnDataSource, Button
from bokeh.plotting import figure

ds = ColumnDataSource(dict(x=[1, 2, 3], y=[1, 2, 3]))
f = IndexFilter(indices=list(range(len(ds.data['x']))))
v = CDSView(source=ds, filters=[f])

fig = figure()
fig.x(x='x', y='y', source=ds, view=v)

b = Button(label='Change indices')


def change_indices():
    if f.indices:
        f.indices = f.indices[:-1]


b.on_click(change_indices)

curdoc().add_root(column(fig, b))
```

Clicking on the button results in `ModelChanged` messages being sent out, but the plot itself does not change.
The root cause is that all filter models do not connect any signals - effectively, they're just passive containers.

As of now, a workaround would be to re-set `filters` property of the `CDSView` instance.

**Comment by bryevdv** [comment]
agree this is an issue of missing plumbing, which might be simple to add

cc @clairetang6

**Comment by birdsarah** [comment]
I just ran into this too when trying to update an IndexFilter. Was extremely confusing.

**Comment by clairetang6** [comment]
The CDSView has a signal set up such that when its `filters` property changes, it will update its indices and send its own change signal to tell the downstream GlyphRenderer to re-render. The issue is that changing an individual filter's properties doesn't cause the CDSView's `filters` property to be "changed", since it is still the same list with the same objects (the same filters) inside. 

There is one approach to fixing this that I like the idea of, but don't know if it's simple to implement. The idea would be to have the filters themselves be "different" when their properties (such as the `indices` in an `IndexFilter` or `booleans` of `BooleanFilter`) have changed. This way, the change detection for the CDSView's `filters` would see the change. In python, this would be easy to implement by giving the filters their own equality definition by overriding `__eq__`, but I don't know if this is a pattern in Javascript/Typescript. 

Another approach to fixing the issue that I like less would be to have the filters send change signals that the CDSView would respond to. The reason I like it less is that that implementation would require the CDSView to keep track of the filters in its `filters` list to connect and disconnect the signals as filters are added or removed.

**Comment by jimbasquiat** [comment]
The solution proposed by p himik seems not to work for datatable:
```
from bokeh.io import curdoc
from bokeh.layouts import column
from bokeh.models import CDSView, IndexFilter, ColumnDataSource, Button, DataTable, TableColumn
from bokeh.plotting import figure

ds = ColumnDataSource(dict(x=[1, 2, 3], y=[1, 2, 3]))
f = IndexFilter(indices=list(range(len(ds.data['x']))))
v = CDSView(source=ds, filters=[f])

columns = [
        TableColumn(field='x', title='x', width = 15),
        TableColumn(field='y', title='y', width = 15)]
data_table = DataTable(
    source = ds, 
    columns = columns,
    view = v,
    width=100, 
    height=100)

b = Button(label='Change indices')

def change_indices():
    if f.indices:
        v.filters = [IndexFilter([0])]

b.on_click(change_indices)

curdoc().add_root(column(data_table, b))
```
(this doesn't update)

It however does work for figure plots: (the following would work):

```
from bokeh.io import curdoc
from bokeh.layouts import column
from bokeh.models import CDSView, IndexFilter, ColumnDataSource, Button, DataTable, TableColumn
from bokeh.plotting import figure

ds = ColumnDataSource(dict(x=[1, 2, 3], y=[1, 2, 3]))
f = IndexFilter(indices=list(range(len(ds.data['x']))))
v = CDSView(source=ds, filters=[f])

fig = figure()
fig.x(x='x', y='y', source=ds, view=v)

b = Button(label='Change indices')

def change_indices():
    if f.indices:
        v.filters = [IndexFilter([0])]


b.on_click(change_indices)

curdoc().add_root(column(fig, b))
```

Any idea why?

**Comment by dkapitan** [comment]
@clairetang6 I have run into this issue as well. Willing to investigate and see if I can come up with a solution. Have read large part of the relevant code, but can't find/figure out where the signals are actually setup. Could you point me in de right direction, as to where to look in the code base?

**Comment by bryevdv** [comment]
Quick note that plumbing is set up to respond to changes of the filter list, just not of changes to existing filters inside the filters list. I.e. , this kind of swap out works:
```
def change_indices():
    f = v.filters[0]
    if f.indices:
        v.filters = [IndexFilter(indices=f.indices[:-1])]
```
bumping this to `next` since it's gotten overlooked

**Comment by bryevdv** [comment]
In retrospect, I rather wish we had simply combined "view" and "filter" since supporting multiple filters per view is a huge complication, as well as nothing anyone has every actually wanted to do in practice in my experience. Maybe we can change course for 3.0 e.g something like

```python
p.scatter('x', 'y', source=s, view=GroupFilterView(group=...))
p.scatter('x', 'y', source=s, view=IndexFilterView(indices=...))
p.scatter('x', 'y', source=s, view=BooleanFilterView(mask=...))
p.scatter('x', 'y', source=s, view=CustomJSFilterView(...))
```

Not having three levels of objects to chain changes together for would be a dramatic simplification, and I think it would satisfy the vast majority of use cases.  cc @bokeh/dev

**Comment by p-himik** [comment]
I would go in a bit different direction. Using both "view" and "filter" in any part of the API makes it much more confusing, so I would just stick to one. If it were up to me, I'd probably choose `filter` because it's more explicit. A view assumes that _anything_ can be done to the data, but it's absolutely not the case - `CDSView` only filters the data.

But it would solve only the most superficial problem with the views/filters. There are other, much more fundamental problems:
- `CDSView` has a **public** state
- A lot of components rely on that state, rely on it being set to just the right values at just the right time, thus coupling things greatly and allowing for a whole family of bugs and inconsistencies (this issue, plus data tables seem to have it as well, judging by the code; I won't bet that there are no other such locations)
- Everything or almost everything that needs to have a data source, now also needs to have a view. The amount of required care has risen dramatically

The proposition - hide all the complexity of the source+view+filter unholy trinity from their clients. Instead, create something like a `FilteredSource` that just wraps the original source and _is_ a proper source for all intents and purposes of the components that end up using it. It will be completely opaque to its users - they will never know that there's even a filter there (of course, one **must not** check `typeof source` - that would negate the whole point). It will still be the same source underneath, all the data changes will still be propagated, all the signals will still (or finally?) be connected automatically just because everything already cares about sources and `FilteredSource` would take care of connecting the filters.
What do you think?

**Comment by bryevdv** [comment]
@p-himik I think we agree 100% on the problems, but I am not so sure about the solution. I do think there is value in having the view be something that is configured on a glyph. It's easy for me to imagine a situation where a user wants to show different views of one data source by adding different views to different glyphs. That's conceptually a model that makes sense to me. But if we couple the view and the filter and the source then all the sudden we have a multiplicity of source types, which I don't like to begin with, but also users would have to create different source objects for different which opens up awhole host of questions about data ownership semantics, etc.

**Comment by p-himik** [comment]
> a situation where a user wants to show different views of one data source by adding different views to different glyphs

And your examples show how this is possible, so it's not an issue. My proposal doesn't break any existing functionality.

> Unlike multiple filters on one glyph, that scenario seems perfectly common and reasonable.

I would argue that multiple filters are also useful. And are even needed, to not make the users reinvent the wheel.
I use them in my projects. One example - a plot with multiple filtering widgets. Each widget controls its own filter. New widgets can be added at runtime to avoid cluttering the UI. I wouldn't want to see this functionality go away and to have redo it all myself in a `CustomJSFilter` or whatever it ends up being.

> all the sudden we have a multiplicity of source types (which I don't like to begin with)

Not of source types but of source wrappers. Right now we have a multiplicity of filters. You offer a multiplicity of views. How is multiplicity of source wrappers different from that?

> `p.scatter('x', 'y', source=s, view=s.group(...))`

Don't do this. This is tight, unnecessary coupling. Now the very basic, core model `ColumnDataSource` knows about all kinds of filters Bokeh users might want.

> `p.scatter('x', 'y', source=GroupSource(s,...))`

This is miles better. It doesn't couple anything - `GroupSource` is just a thin wrapper. `GroupSource(...)` is absolutely not different from `s.group(...)` - it's just a few more characters. And it clearly shows that there's a new model, thus making it more explicit - which is usually better.
Compared to `s.group`, it is also extendable from the outside in a very clear way - it's just a custom new model. A user won't even think about trying to add a new function to the existing `ColumnDataSource` class or to create a new data source class that inherits from CDS. We can even have `MultipleFiltersSource` that has the current behavior of a view with multiple filters.

> what it means to configure one source with another source (which would still be required).

Just to reiterate - I'm proposing wrappers, not fully fledged new sources. They will be sources only according to the LSP ("is a") and not to its actual implementation. (perhaps streaming and patching should be prohibited for such wrappers, so not the truest of "is a")

> users would have to create different source objects for different [glyphs] which opens up awhole host of questions about data ownership semantics, etc

But we already have that problem..? `CDSView` requires `source`, and it must be the same source as the one you set on the glyph. I have no idea if there are checks for that, but I'm pretty sure that providing different sources would be considered an undefined behavior.
So why exactly is the proposed solution more confusing than what users have to do with `CDSView` or with extra filters? If anything, it's _less_ confusing - because one would have to pass only one `source` to the glyphs and other models and not a `source` and a `view` with its own `source` that's actually the same one.

> Everything in this case is: tables and data renderers

There's also `bokeh.models.transforms.Interpolator`. It doesn't know about views now, but ideally it should - interpolate while ignoring outliers is a perfect use case.
There are also `bokeh.models.annotations.LabelSet` and all subclasses of `bokeh.models.annotations.DataAnnotation` - same thing.

With a data source wrapper, everything that needs a source will get automatic support for filtering - the source code of any model that needs a data source won't have to be changed. Including all the possible future models and custom user models - for free! Isn't that neat?

> we now have lots of experience built up to handle the event plumbing for just a single view object

Experience reduces the probability of messing things up. Removing all the places where things can go wrong simply nullifies that probability.
Apart from that, experience is not something that stays with a project - code is.

> Where we got into trouble is eventing for a view, or list of filters on the view, or the filters themselves, or any of the properties on any of the filters, all of which could potentially change at any time or even in principle simultaneously 

And to me, this list seems much longer than it could've been. And it's even a bit shorter than it actually is - you didn't mention the timing coupling: #7299

**Comment by bryevdv** [comment]
> Don't do this. This is tight, unnecessary coupling. Now the very basic, core model ColumnDataSource knows about all kinds of filters Bokeh users might want.

This is purely syntactic sugar:
```
def group(self, **kw): 
    return GroupFilterView(self, **kw)
```
It's an affordance for common cases, nothing more, and does not get int he way of extension filters or customjs filters or just doing things manually if that's what someone prefers.
```
view = GroupFilterView(s, ...)
```

>  it's just a few more characters

It's a few more characters *plus an entire import* and I can't stress enough that Bokeh already gets dinged for being to verbose in this regard, compared to other libraries. 

> Not of source types but of source wrappers. Right now we have a multiplicity of filters. You offer a multiplicity of views. How is multiplicity of source wrappers different from that?

Wrappers are definitely more confusing in my opinion. Many users are not familiar with the concept at all.

* glyph has a source, and optionally a view on that source

is much easier to explain than

* glyph has a source, but that source can be composed via function application to a some facade wrapper that looks like a source, but actually isn't

> But we already have that problem..? CDSView requires source, and it must be the same source as the one you set on the glyph. I have no idea if there are checks for that

There are already validation checks exactly for this. 

> So why exactly is the proposed solution more confusing than what users have to do with CDSView or with extra filters?

Are we talking each other? I don't think your proposed solution is more confusing that the current things users have to do. But I do think encumbering sources is more confusing than maintaining a separation between a source and a view on a source. I have ~10 years of experience having to explain all the various things about CDS, and I very strongly don't want to encumber the "data source" concept itself with *even more* baggage. Furthermore I think the notion of a separate view on to something as its own things is very well established in general. We can rely on that general knowledge to present and explain our particular implementation. 

> There are also bokeh.models.annotations.LabelSet and all subclasses of bokeh.models.annotations.DataAnnotation - same thing.

Those are data renderers.

> And to me, this list seems much longer than it could've been. And it's even a bit shorter than it actually is - you didn't mention the timing coupling: #7299

That's my point! I stopped at a reasonable approximation and it's already too much. Bu my further point is that what I am proposing get's down to waht has already been demonstrated as manageable.

**Comment by p-himik** [comment]
> plus an entire import

I see what you mean, and I guess it's just a detriment of "the Python way" - everywhere you look in the Python documentation, the objects are imported directly and not used via an alias. Nowadays I try to stick to something like this: `import bokeh.models as m`. Then I don't have to change the imports section at all if I need to use something - I just use `m.Model` and whatnot.

> [...]
is much easier to explain than
[...]

I would argue the opposite because that's exactly how Pandas does it, albeit with the same affordances - `iloc` and such are members of a `DataFrame` and they create a view into a data frame. There are no explicit views there at all.

> There are already validation checks exactly for this. Alerting users to mis-configuration is not an issue.

Then why do you think it will create a host of new questions? The situation is exactly the same, only simpler - the same source will have to be mentioned only once per glyph, not twice.

> How does that look in your proposal

As simple as this:
```
ds = ColumnDataSource(...)
p.circle(source=GroupFilterSource(source=ds, column_name='a', group='b'))
[...]
ds.source = new_data
```

The glyph doesn't care - only the wrapper does. It does all the job.
The wrapper source models will simply listen to any changes to `ds` and will recalculate their indices - notifying each and every listener that their `data`, in turn, has changed. And of course their `data` will be not a regular member but a property that returns the data according to the cached indices. Nothing in the outer world will know that it's not the real data and that there's some cache in there.

**Comment by bryevdv** [comment]
you replied while I was editing :) 

> So, technically speaking, the process is exactly the same, just source.data = new_data But now source is some internally configured thing on the wrapper, and not the thing that was configured on the glyph. That is confusing.

I should have added "to me". I'll have to sit with this for some time @p-himik I still don't see the advantages but you clearly feel strongly about it so it deserves more careful consideration. As an aside, this should probably move to a discussion. Unless you object, since there is already considerable discussion here, I'd just convert this issue to a discussion and open a new issue referencing the discussion.

**Comment by p-himik** [comment]
No objections! To be honest, I don't really have a clue whether such a conversion entails something serious, but I'm curious to see how it works out.

**Comment by bryevdv** [comment]
It looks like https://github.com/bokeh/bokeh/issues/11439 is actually a due of this, so I will convert this to a discussion and leave https://github.com/bokeh/bokeh/issues/11439 open to link to eventual PR work.
